### PR TITLE
DS-4283: (Google Scholar) Provide Entity Type Label/Name as part of URL path

### PR DESCRIPTION
**Pull request by benbosman**
## References
_Add references/links to any related issues or PRs. These may include:_
* Part of the REST implementation of https://github.com/DSpace/dspace-angular/issues/745

## Description
This includes:
* The REST API should return the entitytype property
* Sitemaps: The sitemaps currently use the `/items/<:uuid>` syntax. The entity type can either be retrieved from the item metadata or from solr (entity types are already indexed in discovery). The performance impact of either solution would need to be tested first

## Instructions for Reviewers
* The REST API should return the entitytype property
  * You can see the entitytype property is returned if it's present for an item. It is verified in the ItemRestRepositoryIT
* Sitemaps: The sitemaps currently use the `/items/<:uuid>` syntax. The entity type can either be retrieved from the item metadata or from solr (entity types are already indexed in discovery). The performance impact of either solution would need to be tested first
  * The sitemaps will use the format from https://github.com/DSpace/dspace-angular/pull/1001
  * Based on performance tests, the discovery solution has been picked here

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@benbosman : This has a minor code conflict. Could you rebase this when you get the chance?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java [hunk 1]
```diff
@@ -2758,17 +2761,182 @@ public void testHiddenMetadataForUserWithWriteRights() throws Exception {
...
+    }
+
     @Test
     public void findOneTestWithEmbedsWithNoPageSize() throws Exception {
         context.turnOffAuthorisationSystem();
 
         parentCommunity = CommunityBuilder.createCommunity(context)
-                                          .withName("Parent Community")
-                                          .build();
+            .withName("Parent Community")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:2934]: Starting here, your IDE appears to have made a large number of formatting changes.  Could you revert these changes, as it would make this PR a lot smaller & easier to review.... currently over 1/2 the PR is just code realignment.

**Review by tdonohue — APPROVED** [review]
👍 @benbosman : This looks correct & I've tested it and found that it's correctly including `/entities/[type]/` URLs in the Google Sitemaps.  That said, I have one minor requested update before merging this.  **Over 1/2 of this PR appears to be code realignment.  Could you please undo the code realignment in this PR?**

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java [hunk 2]
```diff
@@ -2758,17 +2761,182 @@ public void testHiddenMetadataForUserWithWriteRights() throws Exception {
...
+            .build();
+        context.restoreAuthSystemState();
+
+        String token = getAuthToken(eperson.getEmail(), password);
+
+        getClient(token).perform(get("/api/core/items/" + item.getID()))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$", ItemMatcher.matchItemProperties(item)))
+            .andExpect(jsonPath("$.entityType", is("Project")));
+    }
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:2860]: these tests seem to add no value compared to the first test testEntityTypePerson I would suggest to remove them as they are redundant. It would be useful instead to make test call also with the admin user and as anonymous

File: dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java [hunk 3]
```diff
@@ -237,14 +244,37 @@ public static void generateSitemaps(boolean makeHTMLMap, boolean makeSitemapOrg)
...
             Item i = allItems.next();
-            String url = uiURLStem + "/items/" + i.getID();
-            Date lastMod = i.getLastModified();
 
-            if (makeHTMLMap) {
-                html.addURL(url, lastMod);
-            }
-            if (makeSitemapOrg) {
-                sitemapsOrg.addURL(url, lastMod);
+            DiscoverQuery entityQuery = new DiscoverQuery();
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:248]: if for performance reason we need to make a Discovery query here why not use only discovery/solr to build the sitemaps avoiding the initial items findAll and the single solr query for each items?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ItemConverter.java [hunk 4]
```diff
@@ -51,6 +63,18 @@ public ItemRest convert(Item obj, Projection projection) {
         item.setWithdrawn(obj.isWithdrawn());
         item.setLastModified(obj.getLastModified());
 
+        try {
+            Entity entity = entityService.findByItemId(contextService.getContext(), UUID.fromString(item.getId()));
+            EntityType entityType = entityService.getType(contextService.getContext(), entity);
+            if (entityType != null) {
+                item.setEntityType(entityType.getLabel());
+            }
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ItemConverter.java:71]: due to the performance noted for the sitemaps generator I assume that this operation has an impact, why not simply  return the value of the relationship.type metadata? we know that it represent our entity type label and if later on this rule will change the added test will warn us

**Review by abollini — CHANGES_REQUESTED** [review]
hi @benbosman 
thanks for this PR, I have added inline few suggestion to reduce the performance impact of the needed changes

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java (hunk 2 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:2860]: I would prefer to keep at least tests for 2 different types in place, but we can drop the last 2

Admin and anonymous can be included as well

File: dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java (hunk 3 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:248]: We kept the items findAll to ensure there's no risk of the sitemaps to be incomplete if the search index is being re-built
With the codebase in this PR, it ensures the items are still present, but simply use the /items/uuid syntax if the discovery index is incomplete due to a forced index during the sitemap generation

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ItemConverter.java (hunk 4 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ItemConverter.java:71]: It was done to avoid repeating the relationship.type metadata field name (especially since it's planned to be renamed)
The ITs do indeed indicate it as a problem if this would change, so it can simply return that metadata

File: dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java (hunk 3 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:248]: ok, maybe we can create a backlog issue to further improve that in a future version? we could query the db for the number of items and solr, if they say the same we can use just solr as a reindex is not ongoing for sure.

**Review by abollini — APPROVED** [review]
thanks @benbosman with the changes and explanation provided I'm ok with this PR
@tdonohue please consider this comment https://github.com/DSpace/DSpace/pull/3150#discussion_r592435411

File: dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java (hunk 3 above)
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/sitemap/GenerateSitemaps.java:248]: I went ahead and logged a ticket in the backlog to consider refactoring to use Discovery/Solr in the future: https://github.com/DSpace/DSpace/issues/3182   I agree though that it is low priority at this time.

**Comment by tdonohue** [comment]
This PR was **merged** during a [time when GitHub was having difficulties](https://www.githubstatus.com/incidents/wdhkj70kynwk).  Strangely, this PR did not close after merger, so I'm manually closing it.

However, as you can see it was successfully merged to `main` in commit https://github.com/DSpace/DSpace/commit/c84a349039646ea5b225c1432b182b8896594c35
